src/wasm/wasm-type.cpp

```diff
@@ -390,6 +391,71 @@ bool Type::isRtt() const {
   }
 }
 
+void Type::walk(std::function<void(const Type&)> noteRecursion,
+                std::function<void(const Type&)> visitType) {
+  std::set<Type> scanned;
+  std::vector<Type*> scanList = {this};
+  std::vector<Type*> visitList;
+
+  auto scanHeapType = [&](HeapType& ht) {
+    switch (ht.kind) {
+      case HeapType::FuncKind:
+      case HeapType::ExternKind:
+      case HeapType::ExnKind:
+      case HeapType::AnyKind:
+      case HeapType::EqKind:
+      case HeapType::I31Kind:
+        break;
+      case HeapType::SignatureKind:
+        scanList.push_back(&ht.signature.params);
+        scanList.push_back(&ht.signature.results);
+        break;
+      case HeapType::StructKind:
+        for (auto& field : ht.struct_.fields) {
+          scanList.push_back(&field.type);
+        }
+        break;
+      case HeapType::ArrayKind:
+        scanList.push_back(&ht.array.element.type);
+        break;
+    }
+  };
+
+  // Scan the type tree, noting recursion when we find it
+  while (scanList.size() != 0) {
+    Type* curr = scanList.back();
+    scanList.erase(scanList.end() - 1);
```

Comment:
```suggestion
    scanList.pop_back();
```

Reply:
Isn't pop_back introduced in C++20?

Reply:
No, I think it's been there pretty much all along? https://en.cppreference.com/w/cpp/container/vector/pop_back (looks like c++20 also added a constexpr of it though)

Reply:
Yep, you're totally right, thanks :)